iele-gas.md

```diff
@@ -0,0 +1,960 @@
+ // -----------------------------------------------------------------------
+    rule Cmul(SCHED, L1, L2) => Cmul(SCHED, L2, L1)
+      requires L2 >Int L1
+
+    rule Cmul(SCHED, L1, L2) =>
+        Gmulkara < SCHED > *Int Ckara(L1, L2) +Int
+        Gmulword < SCHED > *Int (L1 +Int L2) +Int
+        Gmul < SCHED >
+      [owise]
+      // Note that if L2 is low enough (< 32) then #overApproxKara(L2) = L2 * L2
+
+    rule Ckara(L1, L2) => L1 *Int #overApproxKara(L2) /Int L2
+      requires L1 >=Int L2
+
+    rule Ckara(L1, L2) => L2 *Int #overApproxKara(L1) /Int L1
+
+    rule Cdiv(SCHED, L1, L2) =>
+        Gdivkara < SCHED > *Int Ckara(L1 -Int L2 +Int 1, L2) +Int
+        Gdivword < SCHED > *Int L1 +Int
+        Gdiv < SCHED >
+      requires L1 >=Int L2
+    
+    rule Cdiv(SCHED, L1, L2) =>
+        Gdivword < SCHED > *Int L1 +Int
+        Gdiv < SCHED >
+      [owise]
+
+    rule Cexp(SCHED, L1, W1, W2) =>
+        Gexpkara < SCHED > *Int #overApproxKara(#adjustedBitLength(L1, W1) *Int W2 /Int 64) +Int
+        Gexpword < SCHED > *Int L1 +Int
+        Gexp < SCHED >
+
+    rule Cexpmod(SCHED, LB, LEX, LM, EX) =>
+        Gexpmodkara < SCHED > *Int #overApproxKara(LM) *Int #adjustedBitLength(LEX, EX) +Int
+        Gexpmod < SCHED >
+      requires LB <=Int LM
+
+    rule Cexpmod(SCHED, LB, LEX, LM, EX) =>
+        Gexpmodkara < SCHED > *Int #overApproxKara(LM) *Int #adjustedBitLength(LEX, EX) +Int
+        Cdiv(SCHED, LB, LM) +Int
+        Gexpmod < SCHED >
+      [owise]
+
+```
+
+#### Approximating `x^log_2 3`
+
+Say we want to approximate `x^log_2 3` with a family of quadratic functions, say of the form `a2*x^2+a1*x+a0`.
+
+* we use the fact that `(2^k)^(log 3 / log 2) = 3^k`
+* then, for `x = 2^k`, it means we want to approximate `3^k` with `a2*4^k(+...)`, whence `a2 ~= 1/(4/3)^k`.
+* looking for powers of 2 (to make `a2*x^2` a shift) which are smaller, but close to `(4/3)^k`, we see that `(4/3)^5 ~= 4.21` and` (4/3)^10 ~=17.75`
+* we then can take candidates `x^2 for x <=32`; `x^2/4 + a1* x + a0 for 32 <= x <=1024`, and  `x^2/16 + b1 * x + b0 for x >= 1024 `
+* now, if we want the approximation to be differentiable, its derivative, `2* x for x <=32; x/2 + a1 for 32 <= x <=1024, x/8 + b1 for x >= 1024`,  must be continuous, so
+* `a1 = 2 * 32 - 32/2 = 48`, and `b1 = 1024/2 + a1 - 1024 / 8 = 432`
+* next, the approximation must also be continuous, so
+* `a0 = 32^2 - 32^2/4 - 48*32 = -768`, and `b0 = 1024^2/4+ 48*1024 -768 - 1024^2/16 - 432*1024 = - 197376`
+
+```{.k .uiuck .rvk}
+    syntax Int ::= #overApproxKara ( Int )                 [function]
+ // -----------------------------------------------------------------
+    rule #overApproxKara(N) => #if N <=Int   32 #then N *Int N
+                         #else #if N <=Int 1024 #then N *Int N /Int  4 +Int  48 *Int N -Int    768
+                         #else                        N *Int N /Int 16 +Int 432 *Int N -Int 197376
```

Comment:
We should make the final case actual behave asymptotically as `x^log_2(3)` or at least a closer overapproximation like `x^1.6`, rather than as `x^2` like this term. What is the upper bound to validity of this last polynomial?